### ISSUE TITLE
OLD class-inheritance concept exercise

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2351]
This issue describes how to implement the `class-inheritance` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/use of `class-inheritance`  in Python. 

## Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes`  in Python
- understanding and use of `single inheritance` in Python
- understanding and use of `multiple inheritance` in Python
  - the diamond inheritance problem
  - Pythons MRO (_method resolution order_)
  - pros and cos of `mixins` () to isolate `multiple inheritance` issues
- when and when not to use `__super__()` to call a parent `class` constructor

  

## Out of scope

- `class-composition` as an explicit topic (_it is ok to use limited composition as needed in example code_)
- `dataclasses`
- `decorators` outside of `@property`, `@staticmethod`, and `@classmethod` (_the student should already be exposed to these in the prerequisite exercise_)
- `generators`
- `coroutines`
- `descriptors` (_these will get their own exercise_)
- using a `class` as a decorator
- performance considerations

  

## Concepts

- `classes`
- `inheritance` in `OOP`
- `single inheritance`
- `multiple inheritance`
- `minxins`
- `__super__()`

  

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `booleans`
- `classes`
- `class-customization`
- `comparisons`
- `decorators`
- `dicts`
- `functions`
- `higher-order-functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

  

## Resources to refer to

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)
- [Real Python: Inheritance and Composition: A Python OOP Guide](https://realpython.com/inheritance-composition-python/)
- [Towards DataScience: Understand Inheritance in Python](https://towardsdatascience.com/understand-inheritance-in-python-74f8e3025f3c) -- (_this is not a good resource for hints or links, since it requires an account signup_)
  - [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
- [Multiple Inheritance in Python](https://www.python-course.eu/python3_multiple_inheritance.php)
- [Multiple Inheritance is Hard (Ned Batchelder)](https://nedbatchelder.com/blog/201210/multiple_inheritance_is_hard.html)
- [Python.org: MRO in Python 2.3+](https://www.python.org/download/releases/2.3/mro/)
- [Dig into Python super() and MRO](https://gaopinghuang0.github.io/2018/12/29/dig-into-python-super-and-mro)
- [Multiple inheritance and mixin classes in Python](https://www.thedigitalcatonline.com/blog/2020/03/27/mixin-classes-in-python/)
- [Making Python classes more modular using mixins](https://www.residentmar.io/2019/07/07/python-mixins.html)
- [Write composable, reusable Python classes using Mixinsl)](https://dev.to/bikramjeetsingh/write-composable-reusable-python-classes-using-mixins-6lj)
- [Mixins in Python and Ruby Compared](https://andrewbrookins.com/technology/mixins-in-python-and-ruby-compared/)
- [Python Data Model - Python Docs](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  - You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.
  

## Concept Description

  **Concept file/issue**:  #2972

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should closely relate to or summarize/paraphrase  the concept introduction document, but add enough additional information and examples  for the student to complete the tasks outlined in this concept exercise.
  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** : `df9adab4-d428-4774-bfd1-54fefe8c9ae7`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
  Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.
- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
Currently being worked on in issue #2960.

**Comment by BethanyG** [comment]
Closing for re-write.
